### PR TITLE
KF-3693 align CI

### PR DESCRIPTION
**Pull request by beliaev-maksim**
Add reusable workflows from 
https://github.com/canonical/charmed-kubeflow-workflows/tree/all/.github/workflows

You may see that the file is quite long and could be further extracted. Unfortunately, current charms state does not allow it.
We need to have the common set of tests across all the charms, then we will be able to further simplify. However, it is already 81line with logic that should not be changed vs 284 lines of constant updates.

Required across the entire Kubeflow portfolio:
* [x] Lint
* [x] Libs
* [x] Unit
* [ ] Observability
* [ ] Bundle
* [ ] Integration (for all the charms within the repo)

Here is another example PR with similar idea: https://github.com/canonical/kfp-operators/pull/187

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+name: CI
+
+on:
+  pull_request:
+  push:
+    branches:
+    - main
+    - track/**
+
+# kill all previous jobs in the same branch if new commits are pushed
```
Comment (beliaev-maksim) [inline .github/workflows/ci.yaml:10]:
```suggestion
# cancel all previous jobs in the same branch if new commits are pushed
```

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+#          - knative-operator
+#          - knative-serving
+#    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_integration-tests.yaml@all
+#    secrets: inherit
+#    with:
+#      charm-name: ${{ matrix.charm }}
+
+  bundle-tests:
+    name: Bundle Tests
+    needs: quality-checks
```
Comment (beliaev-maksim) [inline .github/workflows/ci.yaml:59]: add dep on units

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_bundle.yaml@all
+
+  observability-tests:
+    name: Observability Tests
+    needs: quality-checks
```
Comment (beliaev-maksim) [inline .github/workflows/ci.yaml:64]: add dep on units

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    needs: [quality-checks, unit-tests, bundle-tests, observability-tests]
+    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_publish.yaml@all
+    secrets: inherit
+
+  publish:
+    # always publish on PR
+    name: Publish
+    if: github.event_name == 'pull_request'
+    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_publish.yaml@all
+    secrets: inherit
```
Comment (beliaev-maksim) [inline .github/workflows/ci.yaml:80]:
```suggestion
    secrets: inherit

```

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        charm:
+          - knative-eventing
+          - knative-operator
+          - knative-serving
+    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_unit-tests.yaml@all
+    secrets: inherit
+    with:
+      charm-name: ${{ matrix.charm }}
+
+# uncomment when integration tests are added
```
Comment (DnPlas) [inline .github/workflows/ci.yaml:42]: This is something interesting to discuss and it has to do with the way we write tests. In multi-charm repositories, we have the option to add individual integration tests, if you compare kfp-operators and knative-operators you'll find that we don't have a pattern, these tests may be present or not. We should define the strategy and make a standard, then we'd be able to uncomment or remove this section entirely depending on our decision.

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    needs: quality-checks
+    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_bundle.yaml@all
+
+  observability-tests:
+    name: Observability Tests
+    needs: quality-checks
+    uses: canonical/charmed-kubeflow-workflows/.github/workflows/_observability.yaml@all
+
+  publish-on-success:
+    # only publish on push to main or track/** and if all tests pass
```
Comment (DnPlas) [inline .github/workflows/ci.yaml:68]:
```suggestion
    # only publish on push to main or track/** if all tests pass
```

**Pull request closed without merging**